### DISCUSSION TITLE
New release of CTBase, v0.7.0

### DISCUSSION THREAD

**Discussion by ocots** [control-toolbox/control-toolbox.github.io#41] [Announcements]
See what's new [here](https://github.com/control-toolbox/CTBase.jl/releases/tag/v0.7.0).

Still TBD:

- [Minimal tests](https://github.com/control-toolbox/CTBase.jl/issues/67)
- [Improve abstract possibilities](https://github.com/control-toolbox/CTBase.jl/issues/71)
- [Add Poisson brackets](https://github.com/control-toolbox/CTBase.jl/issues/52)
- Improve the documentation and add coverage

**Comment by ocots** [comment]
@PierreMartinon @joseph-gergaud  @jbcaillau @BaptisteCbl Check this out!

**Reply by jbcaillau** [reply]
🙏🏽 thanks @ocots for the release and the recap: excellent! Having a look at the last changes. 

@PierreMartinon @joseph-gergaud Functional & abstract descriptions are more or less stable 🤞🏾now, we count on you to retrieve a direct solve (= our basic tool) on the new version with variables. Everything should be OK from [`nlp_constraints`](https://github.com/control-toolbox/CTBase.jl/blob/main/src/model.jl#L1065) (tested, but not yet heavily tested) but post an issue if not!

Also need to test the last ADNLP release from our colleagues with sparse Jacobians & Hessians and check performance increase 🤞🏾🤞🏾.

**Comment by BaptisteCbl** [comment]
Hello, if I'm not mistaken, with this new version, CTProblems need to be updated as well as CTFlows (which currently require the old CTBase 0.6.0). I can do pull requests for these two.

**Reply by ocots** [reply]
Thanks!

CTFlows must be updated before CTProblems. The scalar vs vectorial usage has been removed and this has to be taken into account.

You can make a try but I know that CTFlows code is not so clear.

You can also continue on CTBase since there are things to continue as differential geometry.

**Reply by BaptisteCbl** [reply]
Ok thanks, I'll finish things I can in CTBase before looking at it.

**Comment by ocots** [comment]
Update [v0.7.1](https://github.com/control-toolbox/CTBase.jl/releases/tag/v0.7.1).

**Comment by joseph-gergaud** [comment]
With Pierre, we have done the modifications of CTDirect (trapeze). We are waiting the CTProbelm update  for testing all the problems and to merge the branch. Here are some remarks / questions
- Is it possible to have a tag :test in CTProblem which would contain all problems for our tests ?
- It is possible for the moment that a user uses  a scalar for the state in the définition of the  state  equation, but a vector for the lagrangian cost  or for the initial condition. Do you think that it's possible  to check this ?
- In the definition of the problem, it is possible to redefine the cost, but if we redefine a constraint (with the same tag), it's not possible. Before, we have to remove the constraint.
- as a user, I think that it's  better to define the final time as free and not  to define it as a variable.

**Reply by jbcaillau** [reply]
Salut @joseph-gergaud 

> With Pierre, we have done the modifications of CTDirect (trapeze). We are waiting the CTProbelm update for testing all the problems and to merge the branch. Here are some remarks / questions

👍🏽 good news!
 
> * Is it possible to have a tag :test in CTProblem which would contain all problems for our tests ?

@ocots ?

> * It is possible for the moment that a user uses  a scalar for the state in the définition of the  state  equation, but a vector for the lagrangian cost  or for the initial condition. Do you think that it's possible  to check this ?

The user could do that; though not so coherent, it is correct julia syntax so it has to be allowed.

> * In the definition of the problem, it is possible to redefine the cost, but if we redefine a constraint (with the same tag), it's not possible. Before, we have to remove the constraint.

I have just changed this: variable (if any), time, state, control, dynamics and objective declaration must be unique. PR to come.

> * as a user, I think that it's  better to define the final time as free and not  to define it as a variable.

Disagree! If it is free, it is a variable (of the optimisation). This point of view allows us to treat free initial and / or final time and other optimisation variables uniformly. What is the point?

**Reply by jbcaillau** [reply]
@joseph-gergaud check [#88](https://github.com/control-toolbox/CTBase.jl/pull/88)

**Comment by jbcaillau** [comment]
@joseph-gergaud @PierreMartinon Actually, there are quite a few other tests that can be done already, apart from `CTProblem`'s, so it would be nice that you merge in the `develop` branch of `CTDirect`.

**Reply by jbcaillau** [reply]
@ocots I’m not completely sure of the « good » way to test in this case: there are two packages = repos, `A.jl` and `B.jl`. If I want to use not the current release of `B` (`using B...`) from my local version of `A` (that I test using `pkg> activate`), can `pkg> dev` towards the appropriate branch of `B.jl`...?

**Reply by jbcaillau** [reply]
EDIT: should be OK as described [here](https://pkgdocs.julialang.org/v1/managing-packages/#developing) and [there](https://pkgdocs.julialang.org/v1/managing-packages/#Managing-Packages)

**Comment by jbcaillau** [comment]
@joseph-gergaud @PierreMartinon Two questions:
- have you used the last version of `ADNLP` with sparse Jacobians & Hessians?
- it is now possible to define Mayer, Lagrange, and also Bolza cost: have you included this?

**Comment by jbcaillau** [comment]
@joseph-gergaud @PierreMartinon Nice job in `CTDirect.jl`, working fine with variables!
```julia
@def ocp4 begin
    tf ∈ R, variable
    t ∈ [ 0, tf ], time
    x ∈ R², state
    u ∈ R, control
    -1 ≤ u(t) ≤ 1
    x(0) == [ 0, 0 ]
    x(tf) == [ 1, 0 ]
    0.1 ≤ tf ≤ 10
    ẋ(t) == [ x₂(t), u(t) ]
    ∫( 1 ) → min
end

sol4 = solve(ocp4, grid_size=100, print_level=5, tol=1e-12)

plot(sol4)
```

Still an issue with scalar / vector variable, check [#42](https://github.com/control-toolbox/CTDirect.jl/issues/42)

![IMG_2136](https://github.com/control-toolbox/control-toolbox.github.io/assets/62183989/75cd8f8f-8f68-4581-9f1e-ddc70cc3fac2)
